R/plot.data_table_threads_benchmark.R

```diff
@@ -16,46 +16,50 @@
 #' plot(benchmarkData)
 #' }
 
 plot.data_table_threads_benchmark <- function(x, ...)
 {
-  df <- x
-  rownames(df) <- NULL
-  df$speedup <- df$medianTime[df$threadCount == 1] / df$medianTime
-
-  setDT(df)
-  maxSpeedup <- df[, .(threadCount = threadCount[which.max(speedup)], speedup = max(speedup)), by = expr]
-  idealSpeedup <- seq(1, getDTthreads())
-  idealSpeedupData <- data.frame(threadCount = 1:getDTthreads(), speedup = idealSpeedup)
-  subOptimalSpeedupData <- data.frame(threadCount = seq(1, getDTthreads(), length.out = getDTthreads()), speedup = seq(1, getDTthreads()/2, length.out = getDTthreads()))
-
-  closestPoints <- data.frame()
-  for(i in unique(df$expr)) {
-    dfSubset <- df[df$expr == i, ]
-    suboptimalSubset <- subOptimalSpeedupData[subOptimalSpeedupData$threadCount %in% dfSubset$threadCount, ]
-    closestPoint <- dfSubset[which.max(dfSubset$speedup - suboptimalSubset$speedup), ]
-    closestPoints <- rbind(closestPoints, closestPoint)
-  }
-
-  ggplot(df, aes(x = threadCount, y = speedup)) +
-    geom_line(aes(linetype = "Measured")) +
-    geom_line(data = idealSpeedupData, aes(x = threadCount, y = speedup, linetype = "Ideal"), color = "black") +
-    geom_line(data = subOptimalSpeedupData, aes(x = threadCount, y = speedup, linetype = "Sub-optimal"), color = "black") +
-    geom_point(data = closestPoints, aes(x = threadCount, y = speedup, shape = "Recommended"), color = "black", size = 2) +
-    geom_point(data = maxSpeedup, aes(x = threadCount, y = speedup, shape = "Best performing"), color = "red", size = 2) +
-    geom_text(data = closestPoints, aes(label = threadCount), vjust = -0.5, size = 4, na.rm = TRUE) +
-    geom_text(data = maxSpeedup, aes(label = threadCount), vjust = -0.5, size = 4, na.rm = TRUE) +
-    geom_ribbon(aes(ymin = speedup - 0.3, ymax = speedup + 0.3), alpha = 0.5) +
+  x[, `:=`(speedup = medianTime[threadCount == 1] / medianTime, type = "Measured"), by = expr]
+
+  systemThreadCount <- getDTthreads()
+
+  speedupData <- data.table(expr = unique(x$expr))
+  speedupData <- speedupData[, .(
+    threadCount = c(1:systemThreadCount, seq(1, systemThreadCount, length.out = systemThreadCount)),
+    speedup = c(seq(1, systemThreadCount), seq(1, systemThreadCount / 2, length.out = systemThreadCount)),
+    type = rep(c("Ideal", "Recommended"), each = systemThreadCount)
+  ), by = expr]
+
+  maxSpeedup <- x[, .(threadCount = threadCount[which.max(speedup)], speedup = max(speedup), type = "Ideal"), by = expr]
+
+  combinedLineData <- rbindlist(list(speedupData, x), use.names = TRUE, fill = TRUE)
+
+  closestPoints <- x[, {
+    recommendedSpeedupSubset <- speedupData[expr == .BY$expr & type == "Recommended"]
+    merged <- .SD[recommendedSpeedupSubset, on = .(threadCount), nomatch = 0L]
+    .SD[which.max(speedup - merged$speedup)]
+  }, by = expr]
+
+  closestPoints[, `:=`(type = "Recommended")]
+
+  combinedPointData <- rbindlist(list(maxSpeedup, closestPoints), use.names = TRUE, fill = TRUE)
```

Comment:
this pattern is unusual, it seems like you should instead do
`rbind(maxSpeedup, closestPoints)`
(make sure they have the same columns in the same order, or if they don't have the same columns then maybe it is better to avoid rbind and just use two different geoms?)

Reply:
I used `rbindlist` since it is more optimized to combine multiple `data.table` objects (compared to `rbind`), and since we are doing things the `data.table` way I thought that would be better (but I can change it back to `rbind` as that is what I initially used before learning this, and yes I'll need to make sure the columns are consistent - and two different geoms sounds like backtracking so let's avoid that)